### ISSUE TITLE
Add `probability` argument to `LinearSVC`

### ISSUE THREAD

**Issue by CameronBieganek** [alan-turing-institute/MLJ.jl#550]
`SVC` has a `probability` argument. It would be nice if `LinearSVC` also had a `probability` argument so that we can, e.g., calculate AUC. I can see that right [here](https://github.com/alan-turing-institute/MLJModels.jl/blob/d4db7dcf0e9a0fd544b57cc7871496980ea660b6/src/LIBSVM.jl#L409) the probability information is thrown away. 😀

**Comment by OkonSamuel** [comment]
yes `MLJModels` implements LIBSVM models as `Deterministic` models. hence `AUC` measure doesn't apply

**Comment by ablaom** [comment]
Yes, if I recall, this was not implemented at the time because making a SVM probabilisitic is expensive and has issues. See, eg, https://github.com/scikit-learn/scikit-learn/blob/40169b94eb0637b8334bea38072294bc32b17a2d/doc/modules/svm.rst#scores-and-probabilities . However,  I don't see any harm in exposing this, but we should have separate models for the probabilistic solvers, and not implement with a hyper parameter. Ordinarily, we make a model probabilistic by default if possible (and you get point estimates by taking mode/mean or whatever) but here the probabilistic predictor is computationally much more expensive, so we didn't.

**Comment by CameronBieganek** [comment]
> making a SVM probabilisitic is expensive and has issues

Ah, right, that's true if you want to return a calibrated probability. I didn't do much research before creating this issue. It turns out what I actually want is just the decision values/scores. The AUC can be calculated just using the raw decision scores---calibrated probabilities are not needed. It would be cool if there were an easy way to request decision scores in `evaluate!` and have those scores be used to calculate AUC.

For now I could make my own model wrapping `LIBSVM.svmtrain` and `LIBSVM.svmpredict` and add a `predict_scores` function and a modified AUC function. Then I think it would work if I called `evaluate!` like this:

```julia
evaluate!(
    machine(MySVC(), X, y),
    resampling=Holdout(),
    operation=predict_score,
    measure=my_auc
)
```

However, one could argue that `SVC` should default to making uncalibrated probabilistic predictions. In fact, if you look at the plot on [this page](https://scikit-learn.org/stable/modules/calibration.html), we can see that the probabilistic predictions from Naive Bayes and Random Forest classifiers are also uncalibrated. So, since Naive Bayes and Random Forest already return uncalibrated probabilistic predictions in MLJ, for consistency `SVC` should also return an uncalibrated probabilistic prediction. Then we wouldn't need `predict_score`, and `auc` would just work. 🙂

**Comment by ablaom** [comment]
Yes, this makes sense to me, thanks. The new model type should probably get a name like `UncalibratedProbabilisticLinearSVC` to emphasize the different behaviour (keeping my comments in https://github.com/alan-turing-institute/MLJ.jl/issues/557#issuecomment-634393704 in mind). 

Would certainly support a PR in this direction.

**Comment by arvganesh** [comment]
Hi @ablaom, I'm interested in working on this issue but don't see the relevant files (namely `LIBSVM.jl`) in the dev branch, what should I do about this? Thanks so much.

**Comment by ablaom** [comment]
> I'm interested in working on this issue

@arvganesh, that would be great. 

 Yes, the files are hard to find.

The implementation of [MLJ's model API](https://alan-turing-institute.github.io/MLJ.jl/dev/quick_start_guide_to_adding_models/) for the LIBSVM.jl packages is in a separate "interface" package, [MLJLIBSVMInterface.jl](https://github.com/alan-turing-institute/MLJLIBSVMInterface.jl). This is where one would create the new `UncalibratedProbabilisticLinearSVC <: Probabilistic` model. You will see there the existing `Deterministic` models.  

Be sure that `predict(model, fitresult, Xnew)` outputs a `UnivariateFiniteArray` object, constructed using the `UnivariateFinite(...)` [constructor](https://alan-turing-institute.github.io/MLJ.jl/dev/adding_models_for_general_use/#MLJBase.UnivariateFinite). All classes need to be tracked, including those which are not "seen" in the training target (but exist nevertheless as levels, in the sense of `CategoricalArrays`). This usually means bundling some extra stuff in the `fitresult`. A good examplar is [XGBoostClassifier](https://github.com/alan-turing-institute/MLJXGBoostInterface.jl/blob/325c22d3875d9b84540138817660aeb9b7f842fc/src/MLJXGBoostInterface.jl#L583).

**Comment by arvganesh** [comment]
Hi @ablaom, I've been working on this issue for the past few days and have learned something interesting about the versions of `LIBSVM` and `LIBLINEAR`. 

So currently,  I am trying to pass in `probability_estimates=true` to `LIBSVM.LIBLINEAR.linear_predict` defined [here](https://github.com/innerlee/LIBLINEAR.jl/blob/1ce05720f6bd075460b7dfd3474fa4f1507e16f3/src/LIBLINEAR.jl#L248) for `LIBSVM 0.6.0 `and `LIBLINEAR 0.6.0.` 

However, I realized that `MLJLIBSVMInterface.jl` is only meant to be [compatible with](https://github.com/alan-turing-institute/MLJLIBSVMInterface.jl/blob/16bd37d3e20ddad807ddb46023aa3d526c775593/Project.toml#L13) `LIBSVM` of `0.4.0` and `0.5.0`. I downloaded these two versions [manually](https://github.com/JuliaML/LIBSVM.jl/releases) and checked their respective versions of `LIBLINEAR` and found that they both use `LIBLINEAR 0.5.0`.

The relevant part of `linear_predict()` for `LIBLINEAR 0.5.0` (and also `0.5.1`) looks like this:
```julia
fn = probability_estimates ? predict_probability() :
        predict_values()
for i = 1:ninstances
    output = ccall(fn, Float64, (Ptr{Cvoid}, Ptr{FeatureNode}, Ptr{Float64}),
        pointer(m), nodeptrs[i], pointer(decvalues, w_number*(i-1)+1))
    class[i] = model.labels[round(Int,output)]
end
```
The main issue with this (which was [fixed in `LIBLINEAR 0.6.0`](https://github.com/innerlee/LIBLINEAR.jl/compare/v0.5.1...v0.6.0?diff=split#diff-10f297c41f7c15c99f9f2729a223b4418c33d3c9b2f3c4970abccadca65f1581), see `linear_predict()`), is that when output is a probability,`round(Int, output)` can to go `0` and access an invalid index:

When running this: (`MLJLIBSVMInterface v0.1.2`, `LIBSVM 0.5.0`, and `LIBLINEAR 0.5.1`)
```julia
(p, d) = LIBSVM.LIBLINEAR.linear_predict(result, MMI.matrix(Xnew)', probability_estimates=true, verbose=true)
```
I get this issue:
```julia
ERROR: BoundsError: attempt to access 3-element Array{UInt8,1} at index [0]
[1] getindex at ./array.jl:809 [inlined]
[2] linear_predict(::LIBLINEAR.LinearModel{UInt8}, ::LinearAlgebra.Adjoint{Float64,Array{Float64,2}}; probability_estimates::Bool, verbose::Bool) at /Users/arvganesh/.julia/packages/LIBLINEAR/yTdp5/src/LIBLINEAR.jl:334
[3] predict(::UncalibratedProbabilisticLinearSVC, ::Tuple{LIBLINEAR.LinearModel{UInt8},MLJBase.CategoricalDecoder{String,UInt8},CategoricalArrays.CategoricalValue{String,UInt8}}, ::DataFrames.DataFrame) at /Users/arvganesh/Documents/GSOC21/MLJLIBSVMInterface.jl/src/MLJLIBSVMInterface.jl:446
[4] top-level scope at REPL[16]:1
```
where `result` is `LIBSVM.LIBLINEAR.linear_train(...)`

To fix this, I was wondering whether we'd have to upgrade the version of `LIBSVM` that we use to `0.6.0` since it's the only version thus far which is [compatible with](https://github.com/JuliaML/LIBSVM.jl/blob/a2085eae18a9cdd96f688cd08d20e7159fb1b42f/Project.toml#L12) `LIBLINEAR 0.6.0`. Let me know your thoughts and I hope this post was understandable.

**Comment by ablaom** [comment]
@arvganesh Thanks for this investigation. Generally we should be supporting the latest version of these packages. So, before proceeding with the new probabilistic models, let's go ahead and bump the [compat]s for MLJLIBSVMInterface. I think its fine to *bump* rather than extend, if you think the issue you have discovered is likely to complicate support for the LIBSVM 0.5. Is this basically your question?

I see there is already a compathelper PR to extend (https://github.com/alan-turing-institute/MLJLIBSVMInterface.jl/pull/7). So, assuming I have your suggestion  correct, you can go ahead and ignore this and make new PR with [compat] `LIBSVM = "0.6"`, which I can promptly merge.
